crates/payload/builder/src/service.rs

```diff
@@ -127,6 +129,31 @@ where
         }
     }
 
+    /// Sends a message to the service to start building a new payload for the given payload
+    /// attributes and returns a future that resolves to the payload.
+    pub fn send_and_resolve_payload(
+        &self,
+        attr: T::PayloadBuilderAttributes,
+    ) -> MaybePayloadFuture<T::BuiltPayload> {
+        let rx = self.send_new_payload(attr);
+        let to_service = self.to_service.clone();
+
+        Box::pin(async move {
+            let id = match rx.await {
+                Err(e) => return Some(Err(e.into())),
+                Ok(Err(e)) => return Some(Err(e)),
+                Ok(Ok(id)) => id,
+            };
+
+            let (tx, rx) = oneshot::channel();
+            to_service.send(PayloadServiceCommand::Resolve(id, tx)).ok()?;
+            match rx.await.transpose()? {
+                Ok(fut) => Some(fut.await),
+                Err(e) => Some(Err(e.into())),
+            }
```

Comment:
then you can do all of this without the Box::pin

Reply:
Ah ok I thought you wanted to return a future, will update

Reply:
yeah but you can return Option<PayloadFuture>